File: DonRyu/2022-10-04-Coin-Change.js
// 모르겠다
var coinChange = function(coins, amount) {
  var dp = [];
  return coin(coins, amount, dp);
};

var coin = function (coins, amount, dp) {
  if (dp[amount - 1] !== undefined) return dp[amount - 1];
  if (amount < 0) return -1;
  if (amount === 0) return 0;

  var count = Number.MAX_SAFE_INTEGER;
  var tmp = 0;

  for (var i = 0; i < coins.length; i++) {
    tmp = coin(coins, amount - coins[i], dp);
    if (tmp !== -1) count = Math.min(count, tmp + 1);
  }

  dp[amount - 1] = count === Number.MAX_SAFE_INTEGER ? -1 : count;

  return dp[amount - 1];
};

File: SunwooChoi/2022-10-04-Coin-Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        # dp tracks minimum number of coins for amount i
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0
        
        for i in range(1, amount+1):
            for coin in coins:
                if i - coin >= 0:
                    dp[i] = min(dp[i-coin]+1, dp[i])
        
        return -1 if dp[-1] == float('inf') else dp[-1]
